nav2_graceful_motion_controller/src/smooth_control_law.cpp

```diff
@@ -0,0 +1,99 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "nav2_graceful_motion_controller/ego_polar_coords.hpp"
+#include "nav2_graceful_motion_controller/smooth_control_law.hpp"
+
+namespace nav2_graceful_motion_controller
+{
+
+SmoothControlLaw::SmoothControlLaw(
+  double k_phi, double k_delta, double beta, double lambda, double slowdown_radius,
+  double v_linear_min, double v_linear_max, double v_angular_max)
+: k_phi_(k_phi), k_delta_(k_delta), beta_(beta), lambda_(lambda), slowdown_radius_(slowdown_radius),
+  v_linear_min_(v_linear_min), v_linear_max_(v_linear_max), v_angular_max_(v_angular_max)
+{
+}
+
+geometry_msgs::msg::Twist SmoothControlLaw::calculateRegularVelocity(
+  const geometry_msgs::msg::Pose & target, const geometry_msgs::msg::Pose & current)
+{
+  // Convert the target to polar coordinates
+  auto ego_coords = EgocentricPolarCoordinates(target, current);
+  // Calculate the curvature
+  double curvature = calculateCurvature(ego_coords.r, ego_coords.phi, ego_coords.delta);
+
+  // Adjust the linear velocity as a function of the path curvature to
+  // slowdown the controller as it approaches its target
+  double v = v_linear_max_ / (1.0 + beta_ * std::pow(fabs(curvature), lambda_));
+
+  // Set some small v_min when far away from origin to promote faster
+  // turning motion when the curvature is very high
+  v = std::clamp(v, v_linear_min_, v_linear_max_);
+
+  // Slowdown when the robot is near the target to remove singularity
+  v = std::min(v_linear_max_ * (ego_coords.r / slowdown_radius_), v);
+
+  // Compute the angular velocity
+  double w = curvature * v;
+  // Bound angular velocity between [-max_angular_vel, max_angular_vel]
+  double w_bound = std::clamp(w, -v_angular_max_, v_angular_max_);
+  // And linear velocity to follow the curvature
+  v = (curvature != 0.0) ? (w_bound / curvature) : v;
+
+  // Return the velocity command
+  geometry_msgs::msg::Twist cmd_vel;
+  cmd_vel.linear.x = v;
+  cmd_vel.angular.z = w_bound;
+  return cmd_vel;
+}
+
+void SmoothControlLaw::setSpeedLimit(
+  const double v_linear_min, const double v_linear_max,
+  const double v_angular_max)
+{
+  v_linear_min_ = v_linear_min;
+  v_linear_max_ = v_linear_max;
+  v_angular_max_ = v_angular_max;
+}
+
+geometry_msgs::msg::Pose SmoothControlLaw::calculateNextPose(
+  const double dt,
+  const geometry_msgs::msg::Pose & target,
+  const geometry_msgs::msg::Pose & current)
+{
+  geometry_msgs::msg::Twist vel = calculateRegularVelocity(target, current);
+  geometry_msgs::msg::Pose next;
+  double yaw = tf2::getYaw(current.orientation);
+  next.position.x = current.position.x + vel.linear.x * dt * cos(yaw);
+  next.position.y = current.position.y + vel.linear.x * dt * sin(yaw);
+  yaw += vel.angular.z * dt;
+  next.orientation.z = sin(yaw / 2.0);
```

Comment:
I'd generally prefer using TF2 utils to convert from RPY to quaternion than do the conversion here.